src/c2goto/library/pthread_lib.c

```diff
@@ -486,20 +501,37 @@ __ESBMC_HIDE:;
 int pthread_key_create(pthread_key_t *key, void (*destructor)(void *))
 {
 __ESBMC_HIDE:;
-  // TODO
-  return 0;
+  __ESBMC_atomic_begin();
+  __ESBMC_assert(
+    key != NULL,
+    "In pthread_key_create, key parameter should be different than NULL.");
+  __ESBMC_thread_keys[__ESBMC_next_thread_key] = NULL;
+  __ESBMC_thread_key_destructors[__ESBMC_next_thread_key] = destructor;
+  *key = __ESBMC_next_thread_key++;
+  __ESBMC_atomic_end();
+  return nondet_bool() ? (nondet_bool() ? EAGAIN : ENOMEM) : 0;
 }
 
 void *pthread_getspecific(pthread_key_t key)
 {
 __ESBMC_HIDE:;
-  // TODO
-  return NULL;
+  __ESBMC_atomic_begin();
+  if(key > __ESBMC_next_thread_key)
+    return NULL;
+  else
+    return (void *)__ESBMC_thread_keys[key];
+  __ESBMC_atomic_end();
```

Comment:
We are returning without closing the atomic scope, this will "bleed" into other contexts.

Reply:
Fixed.